sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2096,23 +2096,87 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
                 }
-                tableArgs.append(SubqueryAlias(alias, t.evaluable))
-                functionTableSubqueryArgs.append(t)
+                // Check if this is a call to a Python user-defined table function whose polymorphic
+                // 'analyze' method returned metadata indicated requested partitioning and/or
+                // ordering properties of the input relation. In that event, make sure that the UDTF
+                // call did not include any explicit PARTITION BY and/or ORDER BY clauses for the
+                // corresponding TABLE argument, and then update the TABLE argument representation
+                // to apply the requested partitioning and/or ordering.
+                pythonUDTFAnalyzeResult.map { a =>
+                  if (a.withSinglePartition && a.partitionByExpressions.nonEmpty) {
+                    throw QueryCompilationErrors.tableValuedFunctionRequiredMetadataInvalid(
+                      functionName = pythonUDTF.get.name,
+                      reason = "the 'with_single_partition' field cannot be assigned to true " +
+                        "if the 'partition_by' list is non-empty")
+                  } else if (a.orderByExpressions.nonEmpty && !a.withSinglePartition &&
+                    a.partitionByExpressions.isEmpty) {
+                    throw QueryCompilationErrors.tableValuedFunctionRequiredMetadataInvalid(
+                      functionName = pythonUDTF.get.name,
+                      reason = "the 'order_by' field cannot be non-empty unless the " +
+                        "'with_single_partition' field is set to true or the 'partition_by' list " +
+                        "is non-empty")
+                  } else if (a.hasRepartitioning && t.hasRepartitioning) {
+                    throw QueryCompilationErrors
+                      .tableValuedFunctionRequiredMetadataIncompatibleWithCall(
+                        functionName = pythonUDTF.get.name,
+                        requestedMetadata =
+                          "specified its own required partitioning of the input table",
+                        invalidFunctionCallProperty =
+                          "specified the WITH SINGLE PARTITION or PARTITION BY clause; " +
+                            "please remove these clauses and retry the query again.")
+                  }
+                  var withSinglePartition = t.withSinglePartition
+                  var partitionByExpressions = t.partitionByExpressions
+                  var orderByExpressions = t.orderByExpressions
+                  if (a.withSinglePartition) {
+                    withSinglePartition = true
+                  }
+                  if (a.partitionByExpressions.nonEmpty) {
+                    partitionByExpressions = a.partitionByExpressions
+                  }
+                  if (a.orderByExpressions.nonEmpty) {
+                    orderByExpressions = a.orderByExpressions
+                  }
```

Comment:
I'm wondering whether it's ok to just overwrite `orderByExpressions` without any checks.

Reply:
Good question, I originally tried just simply overwriting with the fields of the AnalyzeResult, but found it very easy to make mistakes/bugs by accidentally overwriting details of the UDTF call from the AnalyzeResult when not intended. 

Therefore it seems prudent to explicitly write down every case here in the code, starting with the UDTF call details (from `t`) and then deciding when to overwrite them with the corresponding details of the AnalyzeResult (from `a`). Even if this results in the same behavior for e.g. these `orderByExpressions`, it's probably safer to explicitly write them all out like this in order to make the intentions of the logic clear and to invite consistency later if and when we add more metadata to the UDTF AnalyzeResult.

Reply:
One option to make documentation easier (and make it easier for users to understand), is to simply throw an exception if both the table subquery and the UDTF's analysis result contain any of the metadata.
For example:
UDTFAnalyzeResult: [single_partition = True] is invoked with `udtf(TABLE(t) ORDER BY t.c1) `

We can throw an exception stating that the UDTF already has 'PARTITION BY' or 'ORDER BY' defined in its analysis result, and thus it cannot be used in conjunction with the SQL 'PARTITION BY' or 'ORDER BY' syntax. 

Of course, this approach limits the expressiveness of UDTFs. Another option is to retain the overwriting logic here, but we would need to clearly document the behavior when a UDTF has both an 'ORDER BY' clause in the SQL query and within the UDTF's AnalyzeResult.